### ISSUE TITLE
Request: Eventmessages via OSC

### ISSUE THREAD

**Issue by mcdikki** [CasparCG/Server#343]
As the framing information provided by OSC and INFO command is not reliable (see #107 ),
I would like to suggest a new message type for OSC.

In addition to the messages we're receiving now, we could add an event type of messages that informs the user of interesting changes like a stopped or started producer.
I'm thinking of something like this:

-
**channel/stage/layer/event/load/**
A producer has been loaded to the layer.
```type = {MOVIE | STILL | AUDIO | COLOR}```
```name = {NAME | #COLOR}```
```params = { [LOOP] | [TRANSITION  [DURATION] [TWEEN] [DIRECTION] ] | [SEEK] | [LENGTH] | [FILTER]}```

**channel/stage/layer/event/loadbg/**
A producer has been loaded to the background of the layer.
```type = {MOVIE | STILL | AUDIO | COLOR}```
```name = {NAME | #COLOR}```
```params = { [LOOP] | [TRANSITION  [DURATION] [TWEEN] [DIRECTION] ] | [SEEK] | [LENGTH] | [FILTER] | [AUTO]}```  

**channel/stage/layer/event/play/**
The producer starts playing.
```type = {MOVIE | STILL | AUDIO | COLOR}```
```name = {NAME | #COLOR}```
```params = { [LOOP] | [TRANSITION  [DURATION] [TWEEN] [DIRECTION] ] | [SEEK] | [LENGTH] | [FILTER]}``` 

**channel/stage/layer/event/pause/**
The layer has been paused.

**channel/stage/layer/event/resume/**
The layer has been resumed.

**channel/stage/layer/event/stop/**
The producer of the layer has been stopped.
```type = {MOVIE | STILL | AUDIO | COLOR}```
```name = {NAME | #COLOR}``` 

**channel/stage/layer/event/loop/**
The producer of the layer restarted playing as it was called with the LOOP enabled.
```type = {MOVIE | AUDIO}```
```name = {NAME}```

**channel/stage/layer/event/clear/**
The layer has been cleared.

-
This enables client developers to build better and more reliable clients (especially for playout).
In combination with the idea of a unique id per item as discuses here #121 this would be an extremely powerful feature set.

**Comment by tomtom1976** [comment]
Km

Thomas Petzold (mobil)
TITANIUM media -TITANIUM ggm
sportstiming
Am 01.03.2015 um 17:50 schrieb Christopher pb<notifications@github.com>:

**Comment by mcdikki** [comment]
I'm afraid, but I don't get what you want to tell with this comment Thomas.

**Comment by krzyc** [comment]
Unique identifiers (UUIDs) is a must. Every active object in CasparCG Server should have UUID and there should by possibility to use these identifiers in OSC and AMCP. I have to download 2.1 branch and familiarize with new codebase.

**Comment by ronag** [comment]
I don't quite see why UUIDs would be useful. Every "active object" has a path that should be used as an identifier.

**Comment by ronag** [comment]
OSC should not be used as a stateful API, which events are since package deliver is not guaranteed.

**Comment by mcdikki** [comment]
A stateful wrapper on client side is not possible as long as ccg does not provide reliable framing informations. Please see #107 
I tried this for 2 years now and it is simply not possible. 
There is no client I know off that has reliable, frame accurate way.
This makes ccg good for cg, but unsatisfying for playout. 

Your concerns that OSC over UDP are not the right way of doing event delivery may be right. But at the moment we only have OSC. 
AMCP is a restful API and not an option, too.

If the idea of a unique ID per command as discussed in #121 would be possible, it would be much easier to check own measurements against the real server state. This would enable the clients to workaround the framecount bugs in ccg.  But as long as we only know of the layer state, its not possible to catch all common situations.

**Comment by HellGore** [comment]
Since all OSC messages are sent every frame, there is a good possibility that some (or most) of the UDP packets arrive, therefore a message like
```/channel/1/stage/layer/0/file/frame { 260, 260 }```
will be sent repeatedly at EOF, and the client can react to that (I have not tried to reproduce #107, but of course that should be fixed if it stops at 259).
UPD packets can also be received at the destination out of order, but it is safe to say that the clip will not be playing backwards so that if 260 has been received and 259 arrived later then ignore 259.

To summarize if some extra data is needed, we need to send it every frame (to ensure that clients eventually will have a full state), not one shot events (because they are only sent one time and may get lost). For example we have
```/channel/1/stage/layer/0/paused { true|false }```
being sent every frame which could replace the proposed paused event.

clear is detected by having a time out on the client reacting to the absence of /channel/1/stage/layer/0 messages.

If seamless playlist-playoyt is wanted, the `LOADBG AUTO` is currently the only way.

**Comment by ronag** [comment]
``` clip will not be playing backwards so that if 260 has been received and 259 arrived later then ignore 259.```

Solvable by timestamping each osc packet.

**Comment by ronag** [comment]
```A stateful wrapper on client side is not possible as long as ccg does not provide reliable framing informations.```

And why wouldn't that be possible?

**Comment by mcdikki** [comment]
I accept that sending events via UDP is a bad idea. But then please prioritize fixing #107

>UPD packets can also be received at the destination out of order, but it is safe to say that the clip will not be playing backwards so that if 260 has been received and 259 arrived later then ignore 259.

No, it's not, at least not that simple. A clip could be looping and a currentFrame < prevFrame would be right.
You need the timestamp.

But that isn't the problem. The problem for a stateful wrapper is, that instead of receiving the last frame of a clip, very often you are receiving a framenumber before the last frame several times.
So there is no reliable way of detecting the end of a played clip at client side.

**Comment by dust63** [comment]
Hello,

I'm totally agree with mcdikki. It should be interesting to receive a play/stop osc notification. 
It's not really good to make this process in a wrapper because frame are not really accurrate, and you take some latency to do this.

I think it's a good enhancement for the future version.

Regards.

**Comment by jesperstarkar** [comment]
I imagine events would be a good fit over the 2.1 LOG Socket?
